tls_wrap: migrate synchronous errors

joyeecheung opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

tls_wrap: migrate C++ errors to internal/errors.js

* Throw ERR_TLS_SNI_FROM_SERVER when setting server names on a
  server-side socket instead of returning silently
* Assert on wrap_->started and wrap_->ssl instead of throwing
  errors since these errors indicate that the user either uses
  private APIs, or monkey-patches internals, or hits a bug.

tls_wrap: migrate argument type-checking errors

* Throw ERR_INVALID_ARG_TYPE from public APIs
* Assert argument types in bindings instead of throwing errors

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->
tls, errors

joyeecheung commented:
cc @jasnell @nodejs/tsc 

CI: https://ci.nodejs.org/job/node-test-pull-request/12517/

addaleax commented on lib/_tls_wrap.js:
```diff
@@ -407,7 +410,12 @@ TLSSocket.prototype._wrapHandle = function(wrap) {
-  const res = tls_wrap.wrap(handle._externalStream,
+  const externalStream = handle._externalStream;
+  assert(typeof externalStream === 'object',
+         'handle must be a LibuvStreamWrap');
```
It just needs to be a `StreamBase`, even if `LibuvStreamWrap` is the common case here.

joyeecheung replied:
@addaleax I thought about checking the type of `externalStream` here but there doesn't seem to be a way of checking those in JS land since there isn't a proper inheritance.

addaleax replied:
@joyeecheung Yeah, `externalStream` doesn’t really tell you what kind of thing it is. :/

Ideally, what we could do is making the `StreamBase` inheritance hierarchy reflected in the actual JS handle classes; that requires more work than this PR, though…

joyeecheung replied:
@addaleax Yeah that should probably better be done in a dedicated PR IMO

maclover7 commented on lib/_tls_wrap.js:
```diff
@@ -407,7 +410,12 @@ TLSSocket.prototype._wrapHandle = function(wrap) {
                   tls.createSecureContext(options);
-  const res = tls_wrap.wrap(handle._externalStream,
+  const externalStream = handle._externalStream;
+  assert(typeof externalStream === 'object',
```
Should these be `throw errors.TypeError`s instead of `assert`s?

joyeecheung replied:
@maclover7 I tried that before and the test looked really dubious. These are not user-facing errors, whoever hits then either is using the underscored API or patching internals incorrectly, or is hitting a bug. Either way assertions seem to be more appropriate.

indutny commented:
LGTM

On Fri, Jan 12, 2018 at 7:46 PM Joyee Cheung <notifications@github.com>
wrote:

> *@joyeecheung* commented on this pull request.
> ------------------------------
>
> In lib/_tls_wrap.js
> <https://github.com/nodejs/node/pull/18125#discussion_r161356113>:
>
> > @@ -407,7 +410,12 @@ TLSSocket.prototype._wrapHandle = function(wrap) {
>    const context = options.secureContext ||
>                    options.credentials ||
>                    tls.createSecureContext(options);
> -  const res = tls_wrap.wrap(handle._externalStream,
> +  const externalStream = handle._externalStream;
> +  assert(typeof externalStream === 'object',
>
> @maclover7 <https://github.com/maclover7> I tried that before and the
> test looked really dubious. These are not user-facing errors, whoever hits
> then either is using the underscored API or patching internals incorrectly,
> or is hitting a bug. Either way assertions seem to be more appropriate.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/18125#discussion_r161356113>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAOjw2op6S37uxsZzpAEi3QydHUT1Sjvks5tJ_zlgaJpZM4RczA6>
> .
>

joyeecheung commented:
Updated since https://github.com/nodejs/node/pull/17882 has landed and introduced a check on a changed error. The assertion is probably equally unhelpful as before but since `tls.createSecurePair()` is deprecated I think it's fine not to add another error code for it.

New CI: https://ci.nodejs.org/job/node-test-pull-request/12537/

joyeecheung commented:
CI is green. Can you take a loot again? Thanks! @addaleax @indutny

joyeecheung commented:
Landed in f75bc2c1a5...1c29da8236, thanks!

joyeecheung commented:
q
